syntax = "proto3";

package playlist;

service Playlist {
  rpc GetPlaylist(PlaylistRequest) returns (PlaylistTracksResponse) {}
  rpc CreatePlaylist(CreatePlaylistRequest) returns (PlaylistResponse) {}
  rpc UpdatePlaylist(UpdatePlaylistRequest) returns (PlaylistResponse) {}
  rpc RemovePlaylist(PlaylistRequest) returns (PlaylistResponse) {}

  rpc AddTrackToPlaylist(AddTrackToPlaylistRequest) returns (PlaylistResponse) {}
  rpc RemoveTrackFromPlaylist(AddTrackToPlaylistRequest) returns (PlaylistResponse) {}

  rpc GetPlaylists(PlaylistsRequest) returns (PlaylistsResponse) {}
  rpc GetUserPlaylists(UserPlaylistsRequest) returns (PlaylistsResponse) {}
}

message PlaylistRequest {
  string id = 1;
}

message CreatePlaylistRequest {
  string title = 1;
  optional string thumbnail = 2;
  optional bool is_liked = 3;
}

message UpdatePlaylistRequest {
  string id = 1;
  optional string title = 2;
  optional string thumbnail = 3;
}

message PlaylistResponse {
  string id = 1;
  string title = 2;
  optional string thumbnail = 4;
}

message TrackResponse {
  string id = 1;
}

message PlaylistTracksResponse {
  string id = 1;
  string title = 2;
  optional string thumbnail = 3;
  repeated TrackResponse tracks = 4;
}

message AddTrackToPlaylistRequest {
  string playlist_id = 1;
  string track_id = 2;
}

message PlaylistsRequest {
  uint32 limit = 1;
  uint32 offset = 2;
}

message UserPlaylistsRequest {
  string user_id = 1;
  uint32 limit = 2;
  uint32 offset = 3;
}

message PlaylistsResponse {
  repeated PlaylistResponse playlists = 1;
}
